pkg/cloudprovider/aws/aws.go

```diff
@@ -109,73 +112,145 @@ type ec2InstanceFilter struct {
 }
 
 // True if the passed instance matches the filter
-func (f *ec2InstanceFilter) Matches(instance ec2.Instance) bool {
-	if f.PrivateDNSName != "" && instance.PrivateDNSName != f.PrivateDNSName {
+func (f *ec2InstanceFilter) Matches(instance *ec2.Instance) bool {
+	if f.PrivateDNSName != "" && *instance.PrivateDNSName != f.PrivateDNSName {
 		return false
 	}
 	return true
 }
 
-// goamzEC2 is an implementation of the EC2 interface, backed by goamz
-type goamzEC2 struct {
+// awsSdkEC2 is an implementation of the EC2 interface, backed by aws-sdk-go
+type awsSdkEC2 struct {
 	ec2 *ec2.EC2
 }
 
+func stringPointerArray(orig []string) []*string {
+	n := make([]*string, len(orig))
+	for i, s := range orig {
+		n[i] = &s
+	}
+	return n
+}
+
 // Implementation of EC2.Instances
-func (self *goamzEC2) Instances(instanceIds []string, filter *ec2InstanceFilter) (resp *ec2.InstancesResp, err error) {
-	var goamzFilter *ec2.Filter
-	if filter != nil {
-		goamzFilter = ec2.NewFilter()
-		if filter.PrivateDNSName != "" {
-			goamzFilter.Add("private-dns-name", filter.PrivateDNSName)
+func (self *awsSdkEC2) Instances(instanceIds []string, filter *ec2InstanceFilter) (resp []*ec2.Instance, err error) {
+	var filters []*ec2.Filter
+	if filter != nil && filter.PrivateDNSName != "" {
+		filters = []*ec2.Filter{
+			{
+				Name: aws.String("private-dns-name"),
+				Values: []*string{
+					aws.String(filter.PrivateDNSName),
+				},
+			},
 		}
 	}
-	return self.ec2.Instances(instanceIds, goamzFilter)
+
+	fetchedInstances := []*ec2.Instance{}
+	nextToken := ""
+
+	for {
+		res, err := self.ec2.DescribeInstances(&ec2.DescribeInstancesInput{
+			InstanceIDs: stringPointerArray(instanceIds),
+			Filters:     filters,
+			NextToken:   &nextToken,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		for _, reservation := range res.Reservations {
+			fetchedInstances = append(fetchedInstances, reservation.Instances...)
+		}
+
+		if *res.NextToken == "" {
```

Comment:
I suggest:

```
nextToken = res.NextToken
if isNilOrEmpty(nextToken) {
break
}
```

Where isNilOrEmpty is a helper that returns (s == nil || *s == "")